### PR TITLE
Update `@woocommerce/data` & admin client type definitions

### LINKED ISSUE #32692: Update `@woocommerce/data` types
### Describe the solution you'd like

We should use `@types/wordpress__data` and `@types/wordpress__compose` instead of declaring the modules in the global index.d.ts:

https://github.com/woocommerce/woocommerce/blob/16bf8cb87e4032c105cd5e0e02dd781e73b9ea72/packages/js/data/typings/index.d.ts#L1-L2

Also, some of the data selectors don't export the correct types. It would be great if we could change that.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by chihsuan**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes #32692.

This PR adds the `@types/wordpress__data` and `@types/wordpress__compose`, `redux` types as `@woocommerce/data` devDependencies, and fixes the consequent error.

- [Fix @woocommerce/data type errors after enabling wp data, compose types](https://github.com/woocommerce/woocommerce/pull/32735/commits/5c8f76bbe78db1a219bbb59ac24f90b90a08c771)
- [Fix and export @woocommerce/data types for react admin ./client](https://github.com/woocommerce/woocommerce/pull/32735/commits/672ba17a807962d7eed0fb2597f3dc398d94d5cb)
- [Fix react admin ./client types after updating @woocommerce/data types](https://github.com/woocommerce/woocommerce/pull/32735/commits/678e893f60845385e9b0cbe245379aa29e6e7590)

Many type errors appear because `@types/wordpress__data` is added, making TS able to check related types. This is also why we don't need to manually specify the wp `select` type (https://github.com/woocommerce/woocommerce/pull/32735/commits/d94c1dc7bd0a7c6c7512dda3e4abed9b320ea3b8).

I referred to [wp-calypso](https://github.com/Automattic/wp-calypso/blob/trunk/packages/data-stores/src/site/index.ts#L16-L33) to find out how to add redux store types.

### How to test the changes in this Pull Request:

1. Run `pnpm install`
2. Run `pnpm nx build @woocommerce/data`, confirm no errors
3. Run `pnpm nx ts:check woocommerce-admin`, confirm no errors

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?
* [ ] Have you created a changelog file by running `pnpm nx affected --target=changelog`?

### FOR PR REVIEWER ONLY:

* [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/data/src/onboarding/selectors.ts [hunk 1]
```diff
@@ -204,5 +204,5 @@ export type Extension = {
 	image_url: string;
 	manage_url: string;
 	name: string;
-	slug: string;
+	is_built_by_wc: boolean;
```
Comment (ilyasfoo) [inline packages/js/data/src/onboarding/selectors.ts:207]: Should we also add `is_visible`? https://github.com/woocommerce/woocommerce/blob/0d5a0050d8084a550ef48df76891ab34e4c20b3b/plugins/woocommerce/src/Internal/Admin/RemoteFreeExtensions/DefaultFreeExtensions.php#L83

File: packages/js/data/typings/index.d.ts [hunk 2]
```diff
@@ -1,3 +1,16 @@
...
+declare module '@wordpress/data' {
+	// TODO: update @wordpress/data types to include this.
+	const controls: {
+		select: select;
+		resolveSelect: resolveSelect;
+		dispatch: dispatch;
+	};
+
+
+}
```
Comment (ilyasfoo) [inline packages/js/data/typings/index.d.ts]: Spacing oddities here :D

**Review by ilyasfoo — COMMENTED** [review]
Nice work on this, @chihsuan! I'm leaving feedback but I'm not as confident with my Typescript-fu so I'd prefer someone else in @woocommerce/ghidorah to also review it if needs be.

Feedback: 

1. For all `@ts-expect-error`, do we have a plan to eventually fix them? Or are there any which would not be fixable & would stay indefinitely?

1. Running `pnpm nx ts:check woocommerce-admin` gives me this error (excerpt):

```
...(hundreds more lines)...

client/wp-admin-scripts/payment-method-promotions/payment-promotion-row.tsx:11:8 - error TS7016: Could not find a declaration file for module '@woocommerce/data'. '[redacted]/woocommerce/packages/js/data/build/index.js' implicitly has an 'any' type.
  Try `npm i --save-dev @types/woocommerce__data` if it exists or add a new declaration (.d.ts) file containing `declare module '@woocommerce/data';`

11 } from '@woocommerce/data';
          ~~~~~~~~~~~~~~~~~~~

client/wp-admin-scripts/payment-method-promotions/payment-promotion-row.tsx:55:10 - error TS2339: Property 'installAndActivatePlugins' does not exist on type 'typeof import("[redacted]/woocommerce/node_modules/.pnpm/@types+wordpress__notices@3.3.0/node_modules/@types/wordpress__notices/store/actions")'.

55  const { installAndActivatePlugins } = useDispatch( PLUGINS_STORE_NAME );
            ~~~~~~~~~~~~~~~~~~~~~~~~~

client/wp-admin-scripts/payment-method-promotions/payment-promotion-row.tsx:57:10 - error TS2339: Property 'updatePaymentGateway' does not exist on type 'typeof import("[redacted]/woocommerce/node_modules/.pnpm/@types+wordpress__notices@3.3.0/node_modules/@types/wordpress__notices/store/actions")'.

57  const { updatePaymentGateway } = useDispatch( PAYMENT_GATEWAYS_STORE_NAME );
            ~~~~~~~~~~~~~~~~~~~~

client/wp-admin-scripts/payment-method-promotions/payment-promotion-row.tsx:59:11 - error TS2339: Property 'getPaymentGateway' does not exist on type 'typeof import("[redacted]/woocommerce/node_modules/.pnpm/@types+wordpress__notices@3.3.0/node_modules/@types/wordpress__notices/store/selectors")'.

59   const { getPaymentGateway } = select( PAYMENT_GATEWAYS_STORE_NAME );
             ~~~~~~~~~~~~~~~~~

client/wp-admin-scripts/payment-method-promotions/payment-promotion-row.tsx:62:5 - error TS2339: Property 'getActivePlugins' does not exist on type 'typeof import("[redacted]/woocommerce/node_modules/.pnpm/@types+wordpress__notices@3.3.0/node_modules/@types/wordpress__notices/store/selectors")'.

62   ).getActivePlugins();
       ~~~~~~~~~~~~~~~~


Found 195 errors.

 ELIFECYCLE  Command failed with exit code 1.
```

**Comment by chihsuan** [comment]
@ilyasfoo Thanks for the review!

> For all @ts-expect-error, do we have a plan to eventually fix them? Or are there any which would not be fixable & would stay indefinitely?

Yes. I think our next steps are following the TS migration plan [here](https://github.com/woocommerce/woocommerce/issues/32179) to convert existing JS files to TS. This should be able to fix most of `@ts-expect-error`. For the others, we'll need to update `wordpress/*` type definitions. I'll create a backlog issue for it.

> Running pnpm nx ts:check woocommerce-admin gives me this error (excerpt):

Hmm, this looks like there are no packages type files. Could you try `pnpm nx build:packages woocommerce-admin` and re-run the `pnpm nx ts:check woocommerce-admin` command?

File: packages/js/data/typings/index.d.ts (hunk 2 above)
Comment (chihsuan) [reply packages/js/data/typings/index.d.ts]: 😄 Thanks, reformatted in 022d86c8a588f653df850191e4c21e750e8b635b

File: packages/js/data/src/onboarding/selectors.ts (hunk 1 above)
Comment (chihsuan) [reply packages/js/data/src/onboarding/selectors.ts:207]: Good catch! Added in d0e31f9949d3ec8e6f27bb01cc1b0721b909ed54

**Review by ilyasfoo — COMMENTED** [review]
>Yes. I think our next steps are following the TS migration plan https://github.com/woocommerce/woocommerce/issues/32179 to convert existing JS files to TS. This should be able to fix most of @ts-expect-error. For the others, we'll need to update wordpress/* type definitions. I'll create a backlog issue for it.

Sounds good!

>Hmm, this looks like there are no packages type files. Could you try pnpm nx build:packages woocommerce-admin and re-run the pnpm nx ts:check woocommerce-admin command?

Thanks! This resolved most of the issues but one remains:

```

> nx run woocommerce-admin:"ts:check"

> @woocommerce/admin-library@3.3.0 ts:check /woocommerce/plugins/woocommerce-admin
> tsc --build ./tsconfig.json --pretty

client/tasks/fills/tax/woocommerce-tax/plugins.tsx:71:5 - error TS2578: Unused '@ts-expect-error' directive.

71     // @ts-expect-error PluginInstaller has onComplete props but it is a pure js component and doesn't export the right types.
       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error.

 ELIFECYCLE  Command failed with exit code 1.

———————————————————————————————————————————————

>  NX   ERROR  Running target "woocommerce-admin:ts:check" failed

  Failed tasks:

  - woocommerce-admin:ts:check

  Hint: run the command with --verbose for more details.
```

**Review by adrianduffell — APPROVED** [review]
This tested well and LGTM, thanks @chihsuan  🚀 

Let's merge, though I'll ping @woocommerce/mothra in case they want to check over it too and follow-up with any issues.

**Pull request merged**